sms_analysis.ipynb

```diff
@@ -779,6 +779,149 @@
+    "    \n",
+    "    To determine the \"top\", we look at one cluster's TFIDF - avg(other clusters' TFIDFs)\n",
+    "    \n",
+    "    Args:\n",
+    "        cluster_id: The cluster we want to find the top words for\n",
+    "        x_data_tfidf_sender: TFIDF matrix with as many rows as entries in tfidf_per_sender_cluster\n",
+    "        tfidf_per_sender_cluster: Cluster assignment for each entry in x_data_tfidf_sender\n",
+    "        top_n: Number of top words per cluster to include in the result\n",
+    "    \"\"\"\n",
+    "    this_cluster_indicies = np.argwhere(tfidf_per_sender_cluster == cluster_id)\n",
+    "    other_cluster_indicies = np.argwhere(tfidf_per_sender_cluster != cluster_id)\n",
+    "    this_cluster_records = x_data_tfidf_sender[this_cluster_indicies.flatten()]\n",
+    "    other_cluster_records = x_data_tfidf_sender[other_cluster_indicies.flatten()]\n",
+    "    \n",
+    "    mean_this_cluster = np.asarray(this_cluster_records.mean(axis=0)).reshape(-1)\n",
+    "    mean_other_cluster = np.asarray(other_cluster_records.mean(axis=0)).reshape(-1)\n",
+    "    \n",
+    "    difference = mean_this_cluster - mean_other_cluster\n",
+    "    most_different_indicies = difference.argsort()\n",
+    "    return most_different_indicies[-top_n:]\n",
+    "\n",
+    "def _tfidf_by_sender(slice_of_texts_df, min_texts_required=100):\n",
+    "    \"\"\"Returns a TFIDF matrix of the texts grouped by year.\n",
+    "    \n",
+    "    Years with less than `min_texts_required` texts will be dropped.\n",
+    "    \"\"\"\n",
+    "    #grouper = slice_of_texts_df.date.apply(lambda x: x.full_name)\n",
+    "    grouped_by_name = slice_of_texts_df.groupby(\"full_name\").apply(\n",
+    "        lambda row: pd.Series({'count': len(row.date), 'text': ' '.join(row.text)})\n",
+    "    )\n",
+    "\n",
+    "    # Drops years with less than min_texts_required texts since they won't be very meaningful.\n",
+    "    names_to_drop = grouped_by_name[grouped_by_name['count'] < min_texts_required].index\n",
+    "    grouped_by_name = grouped_by_name[grouped_by_name['count'] >= min_texts_required]\n",
+    "    grouped_by_name.index.name = 'full_name'\n",
+    "\n",
+    "    if grouped_by_name.shape[0] == 0:\n",
+    "        print('Bailing out, no years found with at least {0} texts.'.format(min_texts_required))\n",
+    "        return None\n",
+    "\n",
+    "    grouped_by_name_tfidf = vectorizer.transform(grouped_by_name['text'])\n",
+    "    print('Found {0} conversations with more than {1} texts each.'.format(grouped_by_name_tfidf.shape[0],\n",
+    "                                                                  min_texts_required))\n",
+    "    return grouped_by_name_tfidf, grouped_by_name.index\n",
+    "\n",
+    "tfidf_per_sender, names_sender = _tfidf_by_sender(fully_merged_messages_df[fully_merged_messages_df.is_from_me == 1])\n",
+    "\n",
+    "km_tfidf_sender = KMeans(n_clusters=5)\n",
```

Comment:
Clean up this ordering and naming a bit.  E.g. the flow of why you use SVD twice (once to make kmeans perform better, and once to make it plottable) is not obvious at all.